numpy/core/src/umath/reduction.c

```diff
@@ -537,15 +263,83 @@ PyUFunc_ReduceWrapper(PyArrayObject *operand, PyArrayObject *out,
         op_flags[2] = NPY_ITER_READONLY;
     }
 
+    /* Set up result array axes mapping, operand and wheremask use default */
+    int result_axes[NPY_MAXDIMS];
+    int *op_axes[3] = {result_axes, NULL, NULL};
+
+    int curr_axis = 0;
+    for (int i = 0; i < PyArray_NDIM(operand); i++) {
+        if (axis_flags[i]) {
+            if (keepdims) {
+                result_axes[i] = curr_axis + NPY_ITER_REDUCE_OK;
```

Comment:
This assignment is the one for which my suggestion of setting to `-2` would not work, right?

Reply:
Yeah, its where you need a flag for an existing axis. You could try to not use that (never use keepdims) but that does not seem like a right approach.

In your combination of numbers and flags, what do you mean exactly? Do you mean making `op_axes` longer by inserting flags into it? I do not quite see why newaxis is a flag in this regard.  (Remember this is a mapping `iter_axis→op_axis` and newaxis means it does not exist in the op.)